### PR TITLE
[TASK] Changes on PasswordHashing

### PR DESCRIPTION
**Pull request by franzholz**

### CODE REVIEW

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 1]
```diff
@@ -248,25 +248,26 @@ phpass
...
 
 blowfish
 --------
 
-TYPO3's salted password hash implementation based on `blowfish`_ and PHP`s crypt() function. It has
-been integrated very early to TYPO3 but should no longer be used. It is only included for instances
-that still need to upgrades users to better mechanisms. Options:
+TYPO3's salted password hash implementation based on `blowfish`_ and PHP`s crypt() function. 
+It has been integrated very early to TYPO3 but should no longer be used. It is only included for instances
+that still need to upgrades long time not seen users to better mechanisms. Options:
```
Comment (garvinhicking) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:261]:
```suggestion
that still need to upgrade outdated password hashes to better algorithms. Options:
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 2]
```diff
@@ -248,25 +248,26 @@ phpass
...
 * hash_count: The default log2 number of iterations for password stretching. Defaults to 7.
 
 md5salt
 -------
 
-TYPO3's salted password hash implementation based on `md5`_ and PHP`s crypt() function. It should not be used
-any longer and is only included for instances that still need to upgrade users to better mechanisms.
+TYPO3's salted password hash implementation based on `md5`_ and PHP`s crypt() function.
+It should not be used any longer and is only included for instances that still need 
+to upgrade long time not seen users to better mechanisms.
```
Comment (garvinhicking) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:270]:
```suggestion
to upgrade upgrade outdated password hashes to better algorithms.
```

**Review by garvinhicking — CHANGES_REQUESTED** [review]
Thanks for your contribution! Could you also change the topic of your PR to something like:

```
[TASK] Improve details on PasswordHashing
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 3]
```diff
@@ -95,12 +95,13 @@ a good password hash algorithm and administrators usually do not have to take ca
 is pretty straight forward and helps you to compare passwords with their stored hashes if needed in
 extensions.
 
-One last point on this basic hash knowledge section: Password hashes are always only as secure as
-the user submitted password: If a user has a trivial password like "foo", an attacker who got hold
+One last point on this basic hash knowledge section: Password algorithms are always only as secure as
+the user password: If a user has a trivial password like "foo", an attacker who got hold
```
[the remark this answers is not in the document]
Comment (franzholz) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:99]: It is always about the password submitted to the system. I think it is irrelevant to tell this explicitely.

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 4]
```diff
@@ -130,21 +131,20 @@ Configuration
...
 
 Note that "dead" users (users that don't use the site anymore and never login) will thus never get their
 hashes upgraded to better algorithms. This is an issue that can't be solved on this hash level directly since
 upgrading the password hash always requires the plain text password submitted by the user. However, it is a
-good idea to clean up dead users from the database anyway, site administrators should establish processes to
-comply with the idea of data minimisation of person related data. TYPO3 helps here for instance with the
-"Table garbage collection" task of the scheduler extension, details on this are however out-of-scope of this section.
+good idea to clean up dead users from the database anyway. Site administrators should establish processes to
+comply with the idea of data minimisation of person related data (GDPR). TYPO3 helps here for instance with the
+"Table garbage collection" task of the scheduler extension.
```
[the remark this answers is not in the document]
Comment (franzholz) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:147]: The link must be to this document then.

https://docs.typo3.org/c/typo3/cms-scheduler/main/en-us//Index.html

https://docs.typo3.org/c/typo3/cms-scheduler/main/en-us//Installation/BaseTasks/Index.html?highlight=garbage

[the remark this answers is not in the document]
Comment (franzholz) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: I have added not a link to the garbage collector.

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (garvinhicking) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: You mean you would suggest to drop the whole paragraph?

If we keep it, I'd really like to specify it either more broadly by adressing to it as "a chosen password" or with my sugestion. "User password" sounds oddly referenced, the "submitted" part before the change made the relation imo clearer (but didnt sound that well)

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (garvinhicking) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: Thanks, that helps a lot @brotkrueml can you tell if the notation/link is right, I'm not so well versed in crosslinking yet 😅

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (franzholz) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: It is "chosen password" now.

[the remark this answers is not in the document]
Comment (garvinhicking) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: Thanks, I think that's a good compromise :)

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (brotkrueml) [reply Documentation/ApiOverview/PasswordHashing/Index.rst]: @garvinhicking Syntactically it looks okay. But if you want to be sure, you have to render the PR locally.

**Review by linawolf — APPROVED** [review]
Thanks a lot for these improvements

**Comment by linawolf** [comment]
@brotkrueml can we get this merged?

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 5]
```diff
@@ -96,11 +96,12 @@ is pretty straight forward and helps you to compare passwords with their stored
 extensions.
 
 One last point on this basic hash knowledge section: Password hashes are always only as secure as
-the user submitted password: If a user has a trivial password like "foo", an attacker who got hold
+the chosen password: If a user has a trivial password like "foo", an attacker who has gotten hold
 of the salted password hash will always be successful to crack the hash with a common password hash
 crack tool, no matter how expensive the calculation is. Good password hashing does **not** rescue
-users from short passwords or simple passwords that can be found in a dictionary. It is usually a
-good idea to force users to register with a password that for instance at least has some minimum length.
+users from short or simple passwords that can be found in a dictionary. It is usually a
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:102]: I would rather leave it as is, as the simple passwords is bound to the dictionary, the short password not. Adding a comma makes it clearer:

```suggestion
users from short passwords, or simple passwords that can be found in a dictionary. It is usually a
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 6]
```diff
@@ -96,11 +96,12 @@ is pretty straight forward and helps you to compare passwords with their stored
...
+good idea to force users to register with a password that has for instance at least some minimum length
+and contains even some special characters.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:104]:
```suggestion
and contains even some special characters. See also :ref:`password-policies`.
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 7]
```diff
@@ -130,21 +131,21 @@ Configuration
 Configuration of password hashing is done by TYPO3 automatically and
 administrators usually do not need to worry about details too much: The
 installation process will configure the best available hash algorithm by default.
-This is usually Argon2i. Only if the PHP build is incomplete, some less secure
+This is at the moment Argon2i. Only if the PHP build is incomplete, some less secure
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:134]: That seems to be not really true ("the best hash algorithm"): In v13 the default hashing algorithm will be argon2id:
https://review.typo3.org/c/Packages/TYPO3.CMS/+/82865

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 8]
```diff
@@ -130,21 +131,21 @@ Configuration
...
+configured as default hash algorithm, the user password hash will be upgraded to it. This way, existing user password hashes are updated to newer and better hash algorithms over time, upon login.
 
 Note that "dead" users (users that don't use the site anymore and never login) will thus never get their
 hashes upgraded to better algorithms. This is an issue that can't be solved on this hash level directly since
 upgrading the password hash always requires the plain text password submitted by the user. However, it is a
-good idea to clean up dead users from the database anyway, site administrators should establish processes to
-comply with the idea of data minimisation of person related data. TYPO3 helps here for instance with the
-"Table garbage collection" task of the scheduler extension, details on this are however out-of-scope of this section.
+good idea to clean up dead users from the database anyway. Site administrators should establish processes to
+comply with the idea of data minimisation of person related data (GDPR). TYPO3 helps here for instance with the
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:146]:
```suggestion
comply with the idea of data minimisation of person related data (General Data Protection Regulation, GDPR). TYPO3 helps here for instance with the
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 9]
```diff
@@ -154,7 +155,7 @@ preset of the settings module is available. It can be found in
 .. include:: /Images/AutomaticScreenshots/AdminTools/PasswordHashingSettings.rst.txt
 
 The image shows settings for an instance that runs with frontend and backend users having their passwords
-stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure, they however
+stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure. They however
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:158]: "less secure" -> this does not fit to argon2id.

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 10]
```diff
@@ -213,7 +214,7 @@ Competition in July 2015. There are two available versions:
 Options:
 
 * memory_cost: Maximum memory (in kibibytes) that may be used to compute the Argon2 hash. Defaults to 16384.
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:216]:
```suggestion
* memory_cost: Maximum memory (in kibibytes) that may be used to compute the Argon2 hash. Defaults to 65536.
```

See: https://github.com/TYPO3/typo3/blob/12.4/typo3/sysext/core/Classes/Crypto/PasswordHashing/AbstractArgon2PasswordHash.php#L36

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 11]
```diff
@@ -303,6 +305,8 @@ Example implementation for TYPO3 frontend:
 .. code-block:: php
    :caption: EXT:some_extension/Classes/Controller/SomeController.php
 
+   use TYPO3\CMS\Core\Crypto\PasswordHashing\PasswordHashFactory;
+   ...
```
Comment (brotkrueml) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:309]:
```suggestion
   // ...
```

**Review by brotkrueml — CHANGES_REQUESTED** [review]

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 9 above)
Comment (linawolf) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:158]: How would you rephrase this?

Comment (brotkrueml) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:158]: Maybe something like this: "You should use one of the Argon2 algorithms, as the other listed algorithms ..."

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 7 above)
Comment (linawolf) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:134]: How should we replace it or can we link somewhere where the current default can be determined?

Comment (brotkrueml) [reply Documentation/ApiOverview/PasswordHashing/Index.rst:134]: It is configured in EXT:core/Configuration/DefaultConfiguration.php.

Maybe "one of the best". Argon2i is not the best now.

Background:
https://crypto.stackexchange.com/questions/72416/when-to-use-argon2i-vs-argon2d-vs-argon2id#72437

File: Documentation/ApiOverview/PasswordHashing/Index.rst [hunk 12]
```diff
@@ -154,7 +155,7 @@ preset of the settings module is available. It can be found in
 .. include:: /Images/AutomaticScreenshots/AdminTools/PasswordHashingSettings.rst.txt
 
 The image shows settings for an instance that runs with frontend and backend users having their passwords
-stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure, they however
+stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure. They however
 rely on different PHP capabilities and might be suitable fall backs if Argon2i is not available for whatever
 reason.
```
Comment (linawolf) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:160]:
```suggestion
stored as Argon2i hashes in the database. You should use one of the Argon2 algorithms, as the other listed algorithms are deemed less secure.
They rely on different PHP capabilities and might be suitable fall backs if Argon2i or Argon2id are not available for whatever
reason.
```

File: Documentation/ApiOverview/PasswordHashing/Index.rst (hunk 7 above)
Comment (linawolf) [inline Documentation/ApiOverview/PasswordHashing/Index.rst:134]:
```suggestion
Some of the most secure algorithms currently are Argon2i and Argon2id. Only if the PHP build is incomplete, some less secure
```

**Comment by linawolf** [comment]
@brotkrueml I tried to apply your suggestions. Maybe we can postphone further changes to follow ups? This chapter seems to need a general overhaul, which exeeds the initial intend of this PR

**Review by brotkrueml — APPROVED** [review]

**Pull request merged**